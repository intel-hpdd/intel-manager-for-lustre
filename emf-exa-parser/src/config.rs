// Copyright (c) 2021 DDN. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ExascalerConfiguration {
    pub ha_settings: HaSettings,
    pub global_settings: GlobalSettings,
    pub pool_settings: HashMap<String, PoolSettings>,
    pub sfa_settings: HashMap<String, SfaSettings>,
    pub zpool_settings: Option<HashMap<String, ZpoolSettings>>,
    pub host_defaults_settings: HostDefaultsSettings,
    pub hosts_settings: HashMap<String, HostSettings>,
    pub rest_settings: Option<RestSettings>,
    pub emf_settings: Option<EMFSettings>,
    pub fs_settings: HashMap<String, FsSettings>,
}

impl ExascalerConfiguration {
    /// Return the list of host names in the same group as hostname
    pub fn ha_group_list(&self, host: &str) -> Vec<String> {
        if self.global_settings.clients_list.contains(&host.into()) {
            vec![]
        } else if let Some(settings) = self.hosts_settings.get(host) {
            self.ha_settings.ha_groups[settings.ha_group_idx as usize].clone()
        } else {
            vec![]
        }
    }
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct EMFSettings {
    pub ip: String,
    pub cidr: u32,
    pub size: String,
    pub enabled: bool,
    pub nic: Option<String>,
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct FsSettings {
    pub backfs: String,
    pub cur_mdt_index: u64,
    pub cur_ost_index: u64,
    pub default_ost_count: u64,
    pub default_mdt_count: u64,
    pub dom_enabled: bool,
    pub dom_max_file_size: usize,
    pub host_list: Vec<String>,
    pub hsm_active: bool,
    pub log_dir: String,
    pub mds_list: Vec<String>,
    pub mdt_base_device_path: Option<String>,
    pub mdt_failback: bool,
    pub mdt_list: HashMap<String, Vec<u32>>,
    pub mdt_mke2fs_opts: Option<Vec<String>>,
    pub mdt_mount_opts: Option<String>,
    pub mgs_device_path: Option<String>,
    pub mgs_list: Vec<String>,
    pub mgt_mount_opts: String,
    pub mdt_opts: Option<String>,
    pub mdt_parts: Option<u32>,
    pub mdt_size: String,
    pub mgs_failback: bool,
    pub mgs_internal: bool,
    pub mgs_size: String,
    pub mgs_standalone: bool,
    pub mmp_update_interval: usize,
    pub name: String,
    pub oss_list: Vec<String>,
    pub ost_device_path: Option<String>,
    pub ost_list: HashMap<String, Vec<u32>>,
    pub ost_lnet_list: HashMap<String, Vec<u32>>,
    pub ost_mke2fs_opts: Option<Vec<String>>,
    pub ost_mount_opts: Option<String>,
    pub ost_opts: Option<String>,
    pub pools: Vec<String>,
    pub total_mdt_count: u32,
    pub total_ost_count: u32,
    pub tune_ost_mke2fs_opts: Option<Vec<String>>,
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct GlobalSettings {
    pub clients_list: Vec<String>,
    pub cluster_name: Option<String>,
    pub conf_param_tunings: Option<HashMap<String, String>>,
    pub email_list: Vec<String>,
    pub email_domain: Option<String>,
    pub email_relay: Option<String>,
    pub extra_hosts_start: Vec<String>,
    pub extra_hosts_end: Vec<String>,
    pub fs_list: Vec<String>,
    pub host_list: Vec<String>,
    pub hsm_active: bool,
    pub kdump_path: String,
    pub lnet_mr_fault_sensitive: bool,
    pub log_dir: String,
    pub mdt_backup: String,
    pub mdt_backup_dir: String,
    pub mgs_fs: String,
    pub ntp_list: Vec<String>,
    pub password_policy: String,
    pub pingd: bool,
    pub s2a_list: Vec<String>,
    pub s2a_pass: String,
    pub s2a_user: String,
    pub set_param_tunings: Option<HashMap<String, String>>,
    pub sfa_list: Vec<String>,
    pub shadow_conf: bool,
    pub timezone: Option<String>,
    pub vg_activation_list: Vec<String>,
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct HaSettings {
    pub corosync_nics: Vec<String>,
    pub crypto_cipher: String,
    pub crypto_hash: String,
    pub dampen_ifspeed: u64,
    pub dampen_ping: u64,
    pub failover_policy: String,
    pub ha_group_count: u32,
    pub ha_groups: Vec<Vec<String>>,
    pub lustre_start_timeout: u64,
    pub max_messages: u64,
    pub mcastport: u64,
    pub netmtu: u64,
    pub no_quorum_policy: String,
    pub rrp_mode: String,
    pub secauth: String,
    pub stonith_timeout: u64,
    pub start_on_boot: bool,
    pub transport: String,
    #[serde(rename = "type")]
    pub type_field: String,
    pub window_size: u64,
    pub zpool_monitor_timeout: u64,
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct HostDefaultsSettings {
    pub base_ip: HashMap<String, Vec<u32>>,
    pub bonding_mode: Option<String>,
    pub grub_args: Option<String>,
    pub host_sfa_list: Vec<String>,
    pub ipmi_delay: Option<u64>,
    pub ipmi_method: Option<String>,
    pub ipmi_monitor: Option<u64>,
    pub ipmi_power_wait: u64,
    pub lnets: Vec<String>,
    pub modprobe_cfg: Option<String>,
    pub rest_cert_ca: Option<String>,
    pub rest_cert_crl: Option<String>,
    pub rest_cert_server: Option<String>,
    pub rest_cert_server_key: Option<String>,
    pub rest_ext_nic: Option<String>,
    pub rest_int_nic: Option<String>,
    pub rest_keepalived_nic: Option<String>,
    pub rest_primary_nic: Option<String>,
    pub nic_list: Vec<String>,
    pub nics: HashMap<String, Nic>,
    pub ring0: String,
    pub ring1: String,
    pub serial_port: String,
    pub serial_speed: String,
    pub stonith_pass: Option<String>,
    pub stonith_user: Option<String>,
    pub stonith_type: String,
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct HostSettings {
    pub fs_list: Vec<String>,
    pub bonding_mode: Option<String>,
    pub grub_args: Option<String>,
    pub ha_group: Vec<String>,
    pub ha_group_idx: u32,
    pub host_sfa_list: Vec<String>,
    pub ipmi_delay: Option<u64>,
    pub ipmi_method: Option<String>,
    pub ipmi_monitor: Option<u64>,
    pub ipmi_power_wait: u64,
    pub lnet_members: HashMap<String, Vec<Vec<String>>>,
    pub lnet_nics: Vec<String>,
    pub lnets: Vec<String>,
    pub mdt_base_device_paths: HashMap<String, String>,
    pub mdt_list: HashMap<String, Vec<u32>>,
    pub modprobe_cfg: Option<String>,
    pub name: String,
    pub nic_list: Vec<String>,
    pub nics: HashMap<String, Nic>,
    pub peers: Vec<String>,
    pub oid: Option<String>,
    pub ost_device_paths: HashMap<String, String>,
    pub ost_list: HashMap<String, Vec<u32>>,
    pub rest_cert_ca: Option<String>,
    pub rest_cert_crl: Option<String>,
    pub rest_ext_nic: Option<String>,
    pub rest_int_nic: Option<String>,
    pub rest_cert_server: Option<String>,
    pub rest_cert_server_key: Option<String>,
    pub rest_keepalived_nic: Option<String>,
    pub rest_primary_nic: Option<String>,
    pub ring0: String,
    pub ring1: String,
    pub serial_port: String,
    pub serial_speed: String,
    pub stonith_pass: String,
    pub stonith_primary_peers: Vec<String>,
    pub stonith_secondary_peers: Vec<String>,
    pub stonith_type: String,
    pub stonith_user: String,
    pub sysctl: HashMap<String, String>,
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Nic {
    pub cfg: Option<String>,
    pub device: Option<String>,
    pub gateway: Option<String>,
    pub ip: Option<String>,
    pub is_bonded: Option<bool>,
    pub master: Option<String>,
    pub netaddr: Option<String>,
    pub netmask: Option<String>,
    pub slaves: Option<String>,
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct PoolSettings {
    pub name: String,
    pub ost_list: Vec<String>,
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct RestSettings {
    pub auth_ou: String,
    pub ext_mask: String,
    pub ext_vip: String,
    pub ext_vip_fqdn: String,
    pub int_mask: String,
    pub int_vip: String,
    pub ka_vr_id: String,
    pub master_nodes: Vec<String>,
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct SfaSettings {
    pub controllers: Vec<String>,
    pub name: String,
    pub password: String,
    pub user: String,
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ZpoolSettings {
    pub name: String,
    pub opts: String,
    pub vdev_base_path: String,
    pub vdevs: Vec<String>,
}
