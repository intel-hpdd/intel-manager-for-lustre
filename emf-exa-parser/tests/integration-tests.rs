// Copyright (c) 2021 DDN. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

use emf_exa_parser::config::ExascalerConfiguration;
use insta::{assert_json_snapshot, Settings};

#[test]
fn parse_exascaler_conf_from_file_test() {
    let json = include_str!("fixtures/exascaler.json");
    let exa: ExascalerConfiguration = serde_json::from_str(json).unwrap();
    let mut settings = Settings::new();
    settings.set_sort_maps(true);
    settings.bind(|| assert_json_snapshot!(exa));
}

#[test]
fn parse_exascaler_conf_from_file_test_ai400x() {
    let json = include_str!("fixtures/exascaler_ai400x.json");
    let exa: ExascalerConfiguration = serde_json::from_str(json).unwrap();
    let mut settings = Settings::new();
    settings.set_sort_maps(true);
    settings.bind(|| assert_json_snapshot!(exa));
}

#[test]
fn parse_exascaler_conf_from_file_test_zpool() {
    let json = include_str!("fixtures/exascaler_zpool.json");
    let exa: ExascalerConfiguration = serde_json::from_str(json).unwrap();
    let mut settings = Settings::new();
    settings.set_sort_maps(true);
    settings.bind(|| assert_json_snapshot!(exa));
}

#[test]
fn parse_exascaler_conf_from_file_test_bonded() {
    let json = include_str!("fixtures/exascaler_es7k_with_extMDS.json");
    let exa: ExascalerConfiguration = serde_json::from_str(json).unwrap();
    let mut settings = Settings::new();
    settings.set_sort_maps(true);
    settings.bind(|| assert_json_snapshot!(exa));
}
