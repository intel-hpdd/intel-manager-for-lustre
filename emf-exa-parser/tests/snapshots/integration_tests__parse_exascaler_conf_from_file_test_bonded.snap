---
source: emf-exa-parser/tests/integration-tests.rs
expression: exa

---
{
  "ha_settings": {
    "corosync_nics": [
      "ring0",
      "ring1"
    ],
    "crypto_cipher": "aes256",
    "crypto_hash": "sha1",
    "dampen_ifspeed": 20,
    "dampen_ping": 20,
    "failover_policy": "standard",
    "ha_group_count": 1,
    "ha_groups": [
      [
        "mds0",
        "mds1",
        "vm01-es7k01",
        "vm02-es7k01"
      ]
    ],
    "lustre_start_timeout": 450,
    "max_messages": 17,
    "mcastport": 5405,
    "netmtu": 1500,
    "no_quorum_policy": "freeze",
    "rrp_mode": "passive",
    "secauth": "on",
    "stonith_timeout": 60,
    "start_on_boot": true,
    "transport": "unicast",
    "type": "corosync",
    "window_size": 50,
    "zpool_monitor_timeout": 60
  },
  "global_settings": {
    "clients_list": [],
    "cluster_name": null,
    "conf_param_tunings": null,
    "email_list": [],
    "email_domain": null,
    "email_relay": null,
    "extra_hosts_start": [],
    "extra_hosts_end": [],
    "fs_list": [
      "testfs"
    ],
    "host_list": [
      "mds0",
      "mds1",
      "vm01-es7k01",
      "vm02-es7k01"
    ],
    "hsm_active": false,
    "kdump_path": "/scratch/crash",
    "lnet_mr_fault_sensitive": false,
    "log_dir": "/scratch/log",
    "mdt_backup": "none",
    "mdt_backup_dir": "/scratch",
    "mgs_fs": "testfs",
    "ntp_list": [],
    "password_policy": "plain-text",
    "pingd": true,
    "s2a_list": [],
    "s2a_pass": "password",
    "s2a_user": "admin",
    "set_param_tunings": null,
    "sfa_list": [
      "es7k01",
      "sfa12k"
    ],
    "shadow_conf": false,
    "timezone": null,
    "vg_activation_list": [
      "auto"
    ]
  },
  "pool_settings": {},
  "sfa_settings": {
    "es7k01": {
      "controllers": [
        "192.168.0.100",
        "192.168.0.101"
      ],
      "name": "es7k01",
      "password": "user",
      "user": "user"
    },
    "sfa12k": {
      "controllers": [
        "192.168.0.102",
        "192.168.0.103"
      ],
      "name": "sfa12k",
      "password": "user",
      "user": "user"
    }
  },
  "zpool_settings": null,
  "host_defaults_settings": {
    "base_ip": {
      "bond0": [
        172,
        30,
        12,
        5
      ],
      "eth0": [
        192,
        168,
        0,
        14
      ],
      "ipmi": [
        10,
        128,
        15,
        102
      ]
    },
    "bonding_mode": null,
    "grub_args": null,
    "host_sfa_list": [],
    "ipmi_delay": null,
    "ipmi_method": null,
    "ipmi_monitor": null,
    "ipmi_power_wait": 5,
    "lnets": [
      "tcp(bond0)"
    ],
    "modprobe_cfg": null,
    "rest_cert_ca": null,
    "rest_cert_crl": null,
    "rest_cert_server": null,
    "rest_cert_server_key": null,
    "rest_ext_nic": null,
    "rest_int_nic": null,
    "rest_keepalived_nic": null,
    "rest_primary_nic": null,
    "nic_list": [
      "eth0",
      "bond0",
      "eth1",
      "eth2",
      "eth3",
      "eth4",
      "eth5"
    ],
    "nics": {
      "bond0": {
        "cfg": null,
        "device": "bond0",
        "gateway": null,
        "ip": null,
        "is_bonded": null,
        "master": null,
        "netaddr": null,
        "netmask": "255.240.0.0",
        "slaves": null
      },
      "eth0": {
        "cfg": null,
        "device": "eth0",
        "gateway": "192.168.0.1",
        "ip": null,
        "is_bonded": null,
        "master": null,
        "netaddr": null,
        "netmask": "255.255.255.0",
        "slaves": null
      },
      "eth1": {
        "cfg": null,
        "device": "eth1",
        "gateway": null,
        "ip": null,
        "is_bonded": null,
        "master": null,
        "netaddr": null,
        "netmask": null,
        "slaves": null
      },
      "eth2": {
        "cfg": null,
        "device": "eth2",
        "gateway": null,
        "ip": null,
        "is_bonded": null,
        "master": null,
        "netaddr": null,
        "netmask": null,
        "slaves": null
      },
      "eth3": {
        "cfg": "MTU=9000",
        "device": "eth3",
        "gateway": null,
        "ip": null,
        "is_bonded": null,
        "master": null,
        "netaddr": null,
        "netmask": null,
        "slaves": null
      },
      "eth4": {
        "cfg": "MTU=9000",
        "device": "eth4",
        "gateway": null,
        "ip": null,
        "is_bonded": null,
        "master": null,
        "netaddr": null,
        "netmask": null,
        "slaves": null
      },
      "eth5": {
        "cfg": "MTU=9000",
        "device": "eth5",
        "gateway": null,
        "ip": null,
        "is_bonded": null,
        "master": null,
        "netaddr": null,
        "netmask": null,
        "slaves": null
      },
      "ipmi": {
        "cfg": null,
        "device": null,
        "gateway": null,
        "ip": null,
        "is_bonded": null,
        "master": null,
        "netaddr": null,
        "netmask": "255.255.248.0",
        "slaves": null
      }
    },
    "ring0": "bond0",
    "ring1": "eth0",
    "serial_port": "ttyS0",
    "serial_speed": "115200",
    "stonith_pass": "password",
    "stonith_user": "root",
    "stonith_type": "ipmi"
  },
  "hosts_settings": {
    "mds0": {
      "fs_list": [
        "testfs"
      ],
      "bonding_mode": "6",
      "grub_args": null,
      "ha_group": [
        "mds0",
        "mds1",
        "vm01-es7k01",
        "vm02-es7k01"
      ],
      "ha_group_idx": 0,
      "host_sfa_list": [
        "sfa12k"
      ],
      "ipmi_delay": 15,
      "ipmi_method": "onoff",
      "ipmi_monitor": 60,
      "ipmi_power_wait": 5,
      "lnet_members": {
        "tcp": [
          [
            "172.30.12.1"
          ],
          [
            "172.30.12.2"
          ],
          [
            "172.30.12.3"
          ],
          [
            "172.30.12.4"
          ]
        ]
      },
      "lnet_nics": [
        "bond0"
      ],
      "lnets": [
        "tcp(bond0)"
      ],
      "mdt_base_device_paths": {
        "testfs": "/dev/mapper"
      },
      "mdt_list": {
        "testfs": [
          0
        ]
      },
      "modprobe_cfg": null,
      "name": "mds0",
      "nic_list": [
        "eth0",
        "bond0",
        "eth4",
        "eth5"
      ],
      "nics": {
        "bond0": {
          "cfg": null,
          "device": "bond0",
          "gateway": null,
          "ip": "172.30.12.1",
          "is_bonded": true,
          "master": null,
          "netaddr": "172.16.0.0",
          "netmask": "255.240.0.0",
          "slaves": "eth4 eth5"
        },
        "eth0": {
          "cfg": null,
          "device": "eth0",
          "gateway": "192.168.0.1",
          "ip": "192.168.0.10",
          "is_bonded": false,
          "master": null,
          "netaddr": "192.168.0.0",
          "netmask": "255.255.255.0",
          "slaves": null
        },
        "eth4": {
          "cfg": "MTU=9000",
          "device": "eth4",
          "gateway": null,
          "ip": null,
          "is_bonded": false,
          "master": "bond0",
          "netaddr": null,
          "netmask": null,
          "slaves": null
        },
        "eth5": {
          "cfg": "MTU=9000",
          "device": "eth5",
          "gateway": null,
          "ip": null,
          "is_bonded": false,
          "master": "bond0",
          "netaddr": null,
          "netmask": null,
          "slaves": null
        },
        "ipmi": {
          "cfg": null,
          "device": null,
          "gateway": null,
          "ip": "10.128.15.100",
          "is_bonded": false,
          "master": null,
          "netaddr": null,
          "netmask": "255.255.248.0",
          "slaves": null
        }
      },
      "peers": [
        "mds1"
      ],
      "oid": null,
      "ost_device_paths": {
        "testfs": "/dev/ddn"
      },
      "ost_list": {
        "testfs": []
      },
      "rest_cert_ca": null,
      "rest_cert_crl": null,
      "rest_ext_nic": null,
      "rest_int_nic": null,
      "rest_cert_server": null,
      "rest_cert_server_key": null,
      "rest_keepalived_nic": null,
      "rest_primary_nic": null,
      "ring0": "bond0",
      "ring1": "eth0",
      "serial_port": "ttyS0",
      "serial_speed": "115200",
      "stonith_pass": "password",
      "stonith_primary_peers": [],
      "stonith_secondary_peers": [],
      "stonith_type": "ipmi",
      "stonith_user": "root",
      "sysctl": {
        "vm.min_free_kbytes": "2097152"
      }
    },
    "mds1": {
      "fs_list": [
        "testfs"
      ],
      "bonding_mode": "6",
      "grub_args": null,
      "ha_group": [
        "mds0",
        "mds1",
        "vm01-es7k01",
        "vm02-es7k01"
      ],
      "ha_group_idx": 0,
      "host_sfa_list": [
        "sfa12k"
      ],
      "ipmi_delay": 15,
      "ipmi_method": "onoff",
      "ipmi_monitor": 60,
      "ipmi_power_wait": 5,
      "lnet_members": {
        "tcp": [
          [
            "172.30.12.1"
          ],
          [
            "172.30.12.2"
          ],
          [
            "172.30.12.3"
          ],
          [
            "172.30.12.4"
          ]
        ]
      },
      "lnet_nics": [
        "bond0"
      ],
      "lnets": [
        "tcp(bond0)"
      ],
      "mdt_base_device_paths": {
        "testfs": "/dev/mapper"
      },
      "mdt_list": {
        "testfs": []
      },
      "modprobe_cfg": null,
      "name": "mds1",
      "nic_list": [
        "eth0",
        "bond0",
        "eth4",
        "eth5"
      ],
      "nics": {
        "bond0": {
          "cfg": null,
          "device": "bond0",
          "gateway": null,
          "ip": "172.30.12.2",
          "is_bonded": true,
          "master": null,
          "netaddr": "172.16.0.0",
          "netmask": "255.240.0.0",
          "slaves": "eth4 eth5"
        },
        "eth0": {
          "cfg": null,
          "device": "eth0",
          "gateway": "192.168.0.1",
          "ip": "192.168.0.11",
          "is_bonded": false,
          "master": null,
          "netaddr": "192.168.0.0",
          "netmask": "255.255.255.0",
          "slaves": null
        },
        "eth4": {
          "cfg": "MTU=9000",
          "device": "eth4",
          "gateway": null,
          "ip": null,
          "is_bonded": false,
          "master": "bond0",
          "netaddr": null,
          "netmask": null,
          "slaves": null
        },
        "eth5": {
          "cfg": "MTU=9000",
          "device": "eth5",
          "gateway": null,
          "ip": null,
          "is_bonded": false,
          "master": "bond0",
          "netaddr": null,
          "netmask": null,
          "slaves": null
        },
        "ipmi": {
          "cfg": null,
          "device": null,
          "gateway": null,
          "ip": "10.128.15.101",
          "is_bonded": false,
          "master": null,
          "netaddr": null,
          "netmask": "255.255.248.0",
          "slaves": null
        }
      },
      "peers": [
        "mds0"
      ],
      "oid": null,
      "ost_device_paths": {
        "testfs": "/dev/ddn"
      },
      "ost_list": {
        "testfs": []
      },
      "rest_cert_ca": null,
      "rest_cert_crl": null,
      "rest_ext_nic": null,
      "rest_int_nic": null,
      "rest_cert_server": null,
      "rest_cert_server_key": null,
      "rest_keepalived_nic": null,
      "rest_primary_nic": null,
      "ring0": "bond0",
      "ring1": "eth0",
      "serial_port": "ttyS0",
      "serial_speed": "115200",
      "stonith_pass": "password",
      "stonith_primary_peers": [],
      "stonith_secondary_peers": [],
      "stonith_type": "ipmi",
      "stonith_user": "root",
      "sysctl": {
        "vm.min_free_kbytes": "2097152"
      }
    },
    "vm01-es7k01": {
      "fs_list": [
        "testfs"
      ],
      "bonding_mode": "6",
      "grub_args": null,
      "ha_group": [
        "mds0",
        "mds1",
        "vm01-es7k01",
        "vm02-es7k01"
      ],
      "ha_group_idx": 0,
      "host_sfa_list": [
        "es7k01"
      ],
      "ipmi_delay": null,
      "ipmi_method": null,
      "ipmi_monitor": null,
      "ipmi_power_wait": 5,
      "lnet_members": {
        "tcp": [
          [
            "172.30.12.1"
          ],
          [
            "172.30.12.2"
          ],
          [
            "172.30.12.3"
          ],
          [
            "172.30.12.4"
          ]
        ]
      },
      "lnet_nics": [
        "bond0"
      ],
      "lnets": [
        "tcp(bond0)"
      ],
      "mdt_base_device_paths": {
        "testfs": "/dev/mapper"
      },
      "mdt_list": {
        "testfs": []
      },
      "modprobe_cfg": null,
      "name": "vm01-es7k01",
      "nic_list": [
        "eth0",
        "bond0",
        "eth2",
        "eth3"
      ],
      "nics": {
        "bond0": {
          "cfg": null,
          "device": "bond0",
          "gateway": null,
          "ip": "172.30.12.3",
          "is_bonded": true,
          "master": null,
          "netaddr": "172.16.0.0",
          "netmask": "255.240.0.0",
          "slaves": "eth2 eth3"
        },
        "eth0": {
          "cfg": null,
          "device": "eth0",
          "gateway": "192.168.0.1",
          "ip": "192.168.0.12",
          "is_bonded": false,
          "master": null,
          "netaddr": "192.168.0.0",
          "netmask": "255.255.255.0",
          "slaves": null
        },
        "eth2": {
          "cfg": null,
          "device": "eth2",
          "gateway": null,
          "ip": null,
          "is_bonded": false,
          "master": "bond0",
          "netaddr": null,
          "netmask": null,
          "slaves": null
        },
        "eth3": {
          "cfg": "MTU=9000",
          "device": "eth3",
          "gateway": null,
          "ip": null,
          "is_bonded": false,
          "master": "bond0",
          "netaddr": null,
          "netmask": null,
          "slaves": null
        }
      },
      "peers": [
        "vm02-es7k01"
      ],
      "oid": "0x10000000",
      "ost_device_paths": {
        "testfs": "/dev/ddn"
      },
      "ost_list": {
        "testfs": [
          0,
          1,
          2,
          3,
          4,
          5,
          6
        ]
      },
      "rest_cert_ca": null,
      "rest_cert_crl": null,
      "rest_ext_nic": null,
      "rest_int_nic": null,
      "rest_cert_server": null,
      "rest_cert_server_key": null,
      "rest_keepalived_nic": null,
      "rest_primary_nic": null,
      "ring0": "bond0",
      "ring1": "eth0",
      "serial_port": "ttyS0",
      "serial_speed": "115200",
      "stonith_pass": "user",
      "stonith_primary_peers": [],
      "stonith_secondary_peers": [],
      "stonith_type": "sfa_vm",
      "stonith_user": "user",
      "sysctl": {
        "vm.min_free_kbytes": "2097152"
      }
    },
    "vm02-es7k01": {
      "fs_list": [
        "testfs"
      ],
      "bonding_mode": "6",
      "grub_args": null,
      "ha_group": [
        "mds0",
        "mds1",
        "vm01-es7k01",
        "vm02-es7k01"
      ],
      "ha_group_idx": 0,
      "host_sfa_list": [
        "es7k01"
      ],
      "ipmi_delay": null,
      "ipmi_method": null,
      "ipmi_monitor": null,
      "ipmi_power_wait": 5,
      "lnet_members": {
        "tcp": [
          [
            "172.30.12.1"
          ],
          [
            "172.30.12.2"
          ],
          [
            "172.30.12.3"
          ],
          [
            "172.30.12.4"
          ]
        ]
      },
      "lnet_nics": [
        "bond0"
      ],
      "lnets": [
        "tcp(bond0)"
      ],
      "mdt_base_device_paths": {
        "testfs": "/dev/mapper"
      },
      "mdt_list": {
        "testfs": []
      },
      "modprobe_cfg": null,
      "name": "vm02-es7k01",
      "nic_list": [
        "eth0",
        "bond0",
        "eth2",
        "eth3"
      ],
      "nics": {
        "bond0": {
          "cfg": null,
          "device": "bond0",
          "gateway": null,
          "ip": "172.30.12.4",
          "is_bonded": true,
          "master": null,
          "netaddr": "172.16.0.0",
          "netmask": "255.240.0.0",
          "slaves": "eth2 eth3"
        },
        "eth0": {
          "cfg": null,
          "device": "eth0",
          "gateway": "192.168.0.1",
          "ip": "192.168.0.13",
          "is_bonded": false,
          "master": null,
          "netaddr": "192.168.0.0",
          "netmask": "255.255.255.0",
          "slaves": null
        },
        "eth2": {
          "cfg": null,
          "device": "eth2",
          "gateway": null,
          "ip": null,
          "is_bonded": false,
          "master": "bond0",
          "netaddr": null,
          "netmask": null,
          "slaves": null
        },
        "eth3": {
          "cfg": "MTU=9000",
          "device": "eth3",
          "gateway": null,
          "ip": null,
          "is_bonded": false,
          "master": "bond0",
          "netaddr": null,
          "netmask": null,
          "slaves": null
        }
      },
      "peers": [
        "vm01-es7k01"
      ],
      "oid": "0x10008000",
      "ost_device_paths": {
        "testfs": "/dev/ddn"
      },
      "ost_list": {
        "testfs": [
          7,
          8,
          9,
          10,
          11,
          12,
          13
        ]
      },
      "rest_cert_ca": null,
      "rest_cert_crl": null,
      "rest_ext_nic": null,
      "rest_int_nic": null,
      "rest_cert_server": null,
      "rest_cert_server_key": null,
      "rest_keepalived_nic": null,
      "rest_primary_nic": null,
      "ring0": "bond0",
      "ring1": "eth0",
      "serial_port": "ttyS0",
      "serial_speed": "115200",
      "stonith_pass": "user",
      "stonith_primary_peers": [],
      "stonith_secondary_peers": [],
      "stonith_type": "sfa_vm",
      "stonith_user": "user",
      "sysctl": {
        "vm.min_free_kbytes": "2097152"
      }
    }
  },
  "rest_settings": null,
  "emf_settings": null,
  "fs_settings": {
    "testfs": {
      "backfs": "ldiskfs",
      "cur_mdt_index": 1,
      "cur_ost_index": 14,
      "default_ost_count": 7,
      "default_mdt_count": 0,
      "dom_enabled": false,
      "dom_max_file_size": 65536,
      "host_list": [
        "mds0",
        "mds1",
        "vm01-es7k01",
        "vm02-es7k01"
      ],
      "hsm_active": false,
      "log_dir": "/scratch/log/testfs",
      "mds_list": [
        "mds0",
        "mds1"
      ],
      "mdt_base_device_path": "/dev/mapper",
      "mdt_failback": true,
      "mdt_list": {
        "mds0": [
          0
        ],
        "mds1": []
      },
      "mdt_mke2fs_opts": [
        "-E",
        "lazy_itable_init=0,lazy_journal_init=0",
        "-O",
        "project"
      ],
      "mdt_mount_opts": "max_sectors_kb=0",
      "mgs_device_path": null,
      "mgs_list": [
        "mds0",
        "mds1"
      ],
      "mgt_mount_opts": "max_sectors_kb=0",
      "mdt_opts": null,
      "mdt_parts": 2,
      "mdt_size": "90g",
      "mgs_failback": true,
      "mgs_internal": true,
      "mgs_size": "500m",
      "mgs_standalone": true,
      "mmp_update_interval": 5,
      "name": "testfs",
      "oss_list": [
        "vm01-es7k01",
        "vm02-es7k01"
      ],
      "ost_device_path": "/dev/ddn",
      "ost_list": {
        "vm01-es7k01": [
          0,
          1,
          2,
          3,
          4,
          5,
          6
        ],
        "vm02-es7k01": [
          7,
          8,
          9,
          10,
          11,
          12,
          13
        ]
      },
      "ost_lnet_list": {},
      "ost_mke2fs_opts": [
        "-E",
        "lazy_itable_init=0,lazy_journal_init=0",
        "-O",
        "project"
      ],
      "ost_mount_opts": "max_sectors_kb=0",
      "ost_opts": null,
      "pools": [],
      "total_mdt_count": 1,
      "total_ost_count": 14,
      "tune_ost_mke2fs_opts": []
    }
  }
}
